platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/DefaultTestFilter.java

```diff
@@ -74,64 +68,57 @@ private TestFilter addToFilteringSet(Set<String> filter, String className, Strin
     }
 
     @Override
-    public void setFailOnNoMatchingTests(boolean failOnNoMatchingTests) {
-        this.failOnNoMatching = failOnNoMatchingTests;
-    }
-
-    @Override
-    public boolean isFailOnNoMatchingTests() {
-        return failOnNoMatching;
-    }
+    public abstract Property<Boolean> getFailOnNoMatchingTests();
 
     @Override
     @Input
-    public Set<String> getIncludePatterns() {
-        return includeTestNames;
-    }
+    public abstract SetProperty<String> getIncludePatterns();
 
     @Override
-    public Set<String> getExcludePatterns() {
-        return excludeTestNames;
-    }
+    public abstract SetProperty<String> getExcludePatterns();
 
-    @Override
-    public TestFilter setIncludePatterns(String... testNamePatterns) {
-        return setFilteringPatterns(includeTestNames, testNamePatterns);
+    /**
+     * This is internal property, but it's annotated with @ReplacesEagerProperty too,
+     * since some plugins, e.g. KGP, use it.
+     */
+    @Input
+    @ReplacesEagerProperty(
+        fluentSetter = true,
+        // Kept, since internal classes are not reported in binary checks
+        // so the upgrade check reports an error that original methods was not removed
+        binaryCompatibility = BinaryCompatibility.ACCESSORS_KEPT
+    )
+    public abstract SetProperty<String> getCommandLineIncludePatterns();
+
+    public TestFilter includeCommandLineTest(String className, String methodName) {
+        return addToFilteringSet(getCommandLineIncludePatterns(), className, methodName);
     }
 
-    @Override
-    public TestFilter setExcludePatterns(String... testNamePatterns) {
-        return setFilteringPatterns(excludeTestNames, testNamePatterns);
+    public TestFilterSpec toSpec() {
+        return new TestFilterSpec(
+            // TestFilterSpec get serialized to worker, so we create a copy of original set,
+            // to avoid serialization issues on worker for ImmutableSet that SetProperty returns
+            new LinkedHashSet<>(getIncludePatterns().get()),
+            new LinkedHashSet<>(getExcludePatterns().get()),
+            new LinkedHashSet<>(getCommandLineIncludePatterns().get())
```

Comment:
🤔 This sounds like a leaky abstraction problem. :/ Why wasn't this a problem before?

Reply:
Before we used Java sets which work ok with test workers. But SetProperty.get() returns `ImmutableSet` which we can't serialize to test worker, since it's not on test workers classpath

Reply:
That is going to cause problems for other usages, too. I wonder if this means we need to add support for Groovy immutable types in worker serialization...

Reply:
🤔 Not sure, it might be just a problem for test workers and not other workers, since they have more restricted classpath